### ISSUE TITLE
Add support for `origin` when sending messages

### ISSUE THREAD

**Issue by axe-fb** [react-native-community/react-native-webview#79]
As described in the [reference document](https://github.com/react-native-community/react-native-webview/blob/master/docs/Reference.md), `postMessage` and `onMessage` need to be able to specify the `origin`.

- When a message is received by React Native and then sent to the JS stack, that message should also indicate the URL in the webview where this message came from
- When a message is received by a page in WebView, the origin information should be able to explicitly specify that it is from React Native. 

Alternatively, we could introduce a breaking change, and have a completely separate API to send messages to RN, and to receive messages from React Native

**Comment by Titozzz** [comment]
I think it is effectively a good time to think about a breaking change with a new API, as it has already been done in some libraries. react-native-webview-bridge comes to mind. Relying on post message has many drawbacks and no advantages (that I'm aware of..?).

We need a proper discussion, then we can ask if anyone from the community is willing to implement this. 

What do you think ?

**Comment by jamonholmgren** [comment]
I agree, we should have a solid discussion around this to find a good API.

**Comment by jamonholmgren** [comment]
Related: https://github.com/react-native-community/react-native-webview/issues/29#issuecomment-430779411

**Comment by KoenLav** [comment]
I'm 100% for implementing a breaking change (we could keep a 'backwards compatible' solution by allowing users to enable it with a flag).

**Comment by fungilation** [comment]
I'm for a breaking change in API. Especially if that means sidestepping entirely security issues with `origin`. As a user, it doesn't matter as long as it's documented and works. I've  [advocated for renaming](https://github.com/react-native-community/react-native-webview/issues/66#issue-365173029) so postMessage / onMessage equivalent APIs can work independent of what the WebView's original site javascript does. Sticking with web-centric APIs for RN<>webview communication has never been a wanted feature (for me).

For starter, instead of proposing completely new API, what about just borrow one from existing, working library? [react-native-wkwebview](https://github.com/CRAlpha/react-native-wkwebview)'s API looks comprehensive, but they are using the same `window.postMessage` and `onMessage` API. Do they guarantee against function overwriting from websites' javascript and security issues with origin?

[react-native-webview-bridge](https://github.com/alinz/react-native-webview-bridge) I use actively with my own fork for my own [app](https://medium.com/wonderswipe/rethink-mobile-search-10-100x-faster-introducing-wonderswipe-6f2ff0d0e667). It's not the best API but it works, and it works for both iOS and Android. We could start with that as far as API definition? With one **_exception_**, on calling from RN to WebView:

react-native-webview-bridge uses this to send string to WebView:
```js
webviewbridge.sendToBridge("hello from react-native")
```

but react-native-wkwebview's way is better, since users can directly evaluate JS, and define separate functions for different purposes injected in webview. String sending in react-native-webview-bridge means a nest of conditionals to handle a list of strings sent from RN, and it gets messy. I use react-native-wkwebview so I know.
```js
this.webview.evaluateJavaScript('receivedMessageFromReactNative("Hello from the other side.")')
```

**Comment by Titozzz** [comment]
I'm closing this issue in favor of https://github.com/react-native-community/react-native-webview/issues/66 as it gets super hard to follow more than one thread

**Comment by cnotin** [comment]
The lack of Origin validation is a security issue IMO and it does not seem to have been addressed by #66 (correct me if I'm wrong).
Should not we re-open this?
